Classes/Emogrifier.php

```diff
@@ -537,20 +548,37 @@ private function generateStyleStringFromDeclarationsArrays(array $oldStyles, arr
         return trim($style);
     }
 
-
     /**
      * Copies the media part from CSS array parts to $xmlDocument.
      *
-     * @param string[] $cssParts
-     * @param \DOMDocument $xmlDocument
+     * @param string $mediaCss a string of CSS
+     * @param \DOMDocument $xmlDocument the document to match against
      *
      * @return void
      */
-    public function copyCssWithMediaToStyleNode(array $cssParts, \DOMDocument $xmlDocument)
+    public function copyCssWithMediaToStyleNode($mediaCss, \DOMDocument $xmlDocument)
     {
-        if (isset($cssParts['media']) && $cssParts['media'] !== '') {
-            $this->addStyleElementToDocument($xmlDocument, $cssParts['media']);
+        if ($mediaCss === '') {
+            return;
+        }
+
+        $xpath = new \DOMXPath($xmlDocument);
+        $retainedMediaQueries = [];
+        // Split the CSS into individual media queries.
+        preg_match_all('#(?<query>@media[^{]*\{(?<css>(.*?)\})(\s*)\})#', $mediaCss, $mediaQueries);
+        foreach ($mediaQueries['css'] as $key => $mediaQuery) {
+            // Check if any of the selectors inside the media query apply to the
+            // document.
+            $selectors = $this->parseSelectors($mediaQuery);
+            foreach ($selectors as $selector) {
+                $nodesMatchingSelector = $xpath->query($this->translateCssToXpath($selector['selector']));
+                if ($nodesMatchingSelector && $nodesMatchingSelector->length !== 0) {
+                    $retainedMediaQueries[] = $mediaQueries['query'][$key];
+                }
+            }
         }
+        // Add only the media queries which have relevance for the document.
+        $this->addStyleElementToDocument($xmlDocument, implode('', $retainedMediaQueries));
```

Comment:
If $glue is an empty string, you can also use implode with a single parameter instead.